src/chainstate/stacks/boot/cost-voting.clar

```diff
@@ -0,0 +1,263 @@
+(define-constant REQUIRED_VETOES u500)
+
+;; cost vote token
+(define-fungible-token cost-vote-token)
+
+;; proposal counters
+(define-data-var proposal-count uint u0)
+(define-data-var confirmed-proposal-count uint u0)
+
+;; cost-function proposals
+(define-map proposals
+    { proposal-id: uint }
+    {
+        cost-function-contract: principal,
+        cost-function-name: (string-ascii 128),
+        function-contract: principal,
+        function-name: (string-ascii 128),
+        expiration-block-height: uint
+    }
+)
+
+;; vote confirmed cost-function proposals
+(define-map vote-confirmed-proposals
+    { proposal-id: uint }
+    { expiration-block-height: uint }
+)
+
+;; miner confirmed cost-function proposals
+(define-map confirmed-proposals
+   { confirmed-id: uint }
+   {
+       function-contract: principal,
+       function-name: (string-ascii 128),
+       cost-function-contract: principal,
+       cost-function-name: (string-ascii 128),
+       confirmed-height: uint
+    }
+)
+
+(define-map proposal-confirmed-id
+    { proposal-id: uint }
+    { confirmed-id: uint }
+)
+
+(define-map functions-to-confirmed-ids
+   { function-contract: principal, function-name: (string-ascii 128) }
+   { proposal-id: uint }
+)
+
+;; cost-function proposal votes
+(define-map proposal-votes { proposal-id: uint } { votes: uint })
+
+;; cost-function proposal vetos
+(define-map proposal-vetos { proposal-id: uint } { vetos: uint })
+
+;; proposal vetos per block
+(define-map exercised-veto { proposal-id: uint, veto-height: uint } { vetoed: bool })
+
+;; the number of votes a specific principal has committed to a proposal
+(define-map principal-proposal-votes { address: principal, proposal-id: uint } { votes: uint })
+
+;; getter for cost-function proposals
+(define-read-only (get-proposal (proposal-id uint))
+    (map-get? proposals { proposal-id: proposal-id }))
+
+;; getter for confirmed cost-function proposals
+(define-read-only (get-confirmed-proposal (confirmed-id uint))
+    (map-get? confirmed-proposals { confirmed-id: confirmed-id }))
+
+;; getter for cost-function proposal votes
+(define-read-only (get-proposal-votes (proposal-id uint))
+    (get votes (map-get? proposal-votes { proposal-id: proposal-id })))
+
+;; getter for cost-function proposal vetos
+(define-read-only (get-proposal-vetos (proposal-id uint))
+    (get vetos (map-get? proposal-vetos { proposal-id: proposal-id })))
+
+;; getter for cost-function proposal votes, for specific principal
+(define-read-only (get-principal-votes (address principal) (proposal-id uint))
+    (get votes (map-get? principal-proposal-votes { address: address, proposal-id: proposal-id })))
+
+;; Propose cost-functions
+(define-public (submit-proposal (function-contract principal)
+                                (function-name (string-ascii 128))
+                                (cost-function-contract principal)
+                                (cost-function-name (string-ascii 128)))
+    (begin
```

Comment:
Just want to confirm -- if somehow the chain manages to confirm a cost function that doesn't exist or is malformed (e.g. it does I/O), the cost-assessment logic will fall back to the last valid cost-function for the given function here?

Also, what happens if `function-contract` and `function-name` don't correspond to a real function?  Will the cost function still apply to it even if the function is created later?

Reply:
Those questions are answered in #2143 -- invalid proposals are simply never "applied" to the cost tracker. A warning is emitted, and then they aren't considered again.

Reply:
Thanks!